crates/turbo-tasks-memory/src/scope.rs

```diff
@@ -194,115 +201,101 @@ impl TaskScope {
         self.tasks.fetch_sub(1, Ordering::Relaxed);
     }
 
-    pub fn increment_unfinished_tasks(&self) {
-        self.unfinished_tasks.fetch_add(1, Ordering::AcqRel);
+    pub fn increment_unfinished_tasks(&self, backend: &MemoryBackend) {
+        if self.increment_unfinished_tasks_internal() {
+            self.update_unfinished_state(backend);
+        }
+    }
+
+    /// Returns true if the state requires an update
+    #[must_use]
+    fn increment_unfinished_tasks_internal(&self) -> bool {
+        // crossing to 0 to 1 boundary, requires an update
+        self.unfinished_tasks.fetch_add(1, Ordering::AcqRel) == 0
     }
 
     pub fn decrement_unfinished_tasks(&self, backend: &MemoryBackend) {
+        if self.decrement_unfinished_tasks_internal() {
+            self.update_unfinished_state(backend);
+        }
+    }
+
+    /// Returns true if the state requires an update
+    #[must_use]
+    fn decrement_unfinished_tasks_internal(&self) -> bool {
         let old_count = self.unfinished_tasks.fetch_sub(1, Ordering::AcqRel);
         debug_assert!(old_count > 0);
```

Comment:
Can't we go below 0 now with race conditions?

Reply:
yep, I need to remove that assert...